toqito/perms/permutation_operator.py

```diff
@@ -54,10 +53,10 @@ def permutation_operator(
     """
     # Allow the user to enter a single number for `dim`.
     if isinstance(dim, int):
-        dim = dim * np.ones(max(perm))
+        dim = [dim] * len(perm)
```

Comment:
So the line generates a numpy array filled with the max value of perm 
in this example `res = permutation_operator(2, [1, 3, 2])`, max of perm gives us an dimension array of 3s,  which makes sense.  But now that we have moved to a 0 index system,  for `res = permutation_operator(2, [0, 2, 1])` we would get an array of 2s for dimension when we need an array of 3s.

The change from max to len was made for that reason.
Also, I didn't use np.ones(len(perm)) because there is the if case in the next line that checks if dim is a list and converts it into a numpy array, 
`dim = [dim] * len(perm)` creates a dimension list `[2]*3` which then laters gets updated into an array, I can convert the list to an array in the same line if you wish 
`dim = [dim] * np.ones(len(perm))`

Reply:
I don't think the logic of the code block is working as you expect it to .

I used print statements to see how the `dim` argument is being used here. With your changes, 1 `if` condition creates a list object and the other one created a numpy array. Plus, once you use the first `if`, it shouldn't go into the second `if` considering you have created the `dim` you require but that's what this code block is currently doing. 

```python
    if isinstance(dim, int):
        dim = [dim] * len(perm)
        print(type(dim))
        print("instance is int dim", dim, '\n')
    if isinstance(dim, list):
        dim = np.array(dim)
        print(type(dim))
        print("instance is list dim",dim, '\n')
```

For `permutation_operator(2, [2, 1])`, the output due to the changes is
```
<class 'list'>
instance is int dim [2, 2] 

<class 'numpy.ndarray'>
instance is list dim [2 2] 

[[1. 0. 0. 0.]
 [0. 1. 0. 0.]
 [0. 0. 1. 0.]
 [0. 0. 0. 1.]] 

array([[1., 0., 0., 0.],
       [0., 0., 1., 0.],
       [0., 1., 0., 0.],
       [0., 0., 0., 1.]])
```

whereas you expect it to be 

```
<class 'numpy.ndarray'>
instance is int dim [2. 2.] 

[[1. 0. 0. 0.]
 [0. 1. 0. 0.]
 [0. 0. 1. 0.]
 [0. 0. 0. 1.]] 

array([[1., 0., 0., 0.],
       [0., 0., 1., 0.],
       [0., 1., 0., 0.],
       [0., 0., 0., 1.]])
```

Reply:
I'll add a comment later with an example about `max(perm)` vs. `len(perm)`. 

From what I found, it still doesn't make sense to me. Need to work on simple demo example before resolving this thread.

Reply:
> Plus, once you use the first `if`, it shouldn't go into the second `if` considering you have created the `dim` you require but that's what this code block is currently doing.
> 
> ```python
>     if isinstance(dim, int):
>         dim = [dim] * len(perm)
>         print(type(dim))
>         print("instance is int dim", dim, '\n')
>     if isinstance(dim, list):
>         dim = np.array(dim)
>         print(type(dim))
>         print("instance is list dim",dim, '\n')
> ```
> 
> For `permutation_operator(2, [2, 1])`, the output due to the changes is
> 
> ```
> <class 'list'>
> instance is int dim [2, 2] 
> 
> <class 'numpy.ndarray'>
> instance is list dim [2 2] 
> 
> [[1. 0. 0. 0.]
>  [0. 1. 0. 0.]
>  [0. 0. 1. 0.]
>  [0. 0. 0. 1.]] 
> 
> array([[1., 0., 0., 0.],
>        [0., 0., 1., 0.],
>        [0., 1., 0., 0.],
>        [0., 0., 0., 1.]])
> ```
> 
> whereas you expect it to be
> 
> ```
> <class 'numpy.ndarray'>
> instance is int dim [2. 2.] 
> 
> [[1. 0. 0. 0.]
>  [0. 1. 0. 0.]
>  [0. 0. 1. 0.]
>  [0. 0. 0. 1.]] 
> 
> array([[1., 0., 0., 0.],
>        [0., 0., 1., 0.],
>        [0., 1., 0., 0.],
>        [0., 0., 0., 1.]])
> ```

I do see the issue , yes you are right the first if triggers if dim is an integer and the second if given as a list, which makes sense, will correct it, sorry not sure how I missed this, but regarding the max and len, in 1 index list the max perm is essentially the size of the perm but in 0 index its max+1 rather I changed it to be len which is the ideal one, but I am open to learn why this is incorrect or wont work here

Reply:
> regarding the max and len, in 1 index list the max perm is essentially the size of the perm but in 0 index its max+1 rather I changed it to be len which is the ideal one

I am having some trouble trying to understand what you are saying here. Could you provide a simple example?

`max` finds the maximum value in an array or along an axis whereas `len` gives you the length of something. Where does the issue of shifted indices affect the output below? 

```python
perm = [7, 8, 9, 10]
print(max(perm))
print(len(perm))
```

Reply:
> > regarding the max and len, in 1 index list the max perm is essentially the size of the perm but in 0 index its max+1 rather I changed it to be len which is the ideal one
> 
> I am having some trouble trying to understand what you are saying here. Could you provide a simple example?
> 
> `max` finds the maximum value along an axis (if we choose to use it) whereas `len` gives you the length of something. Where does the issue of shifted indices affect the output below?
> 
> ```python
> perm = [7, 8, 9, 10]
> print(max(perm))
> print(len(perm))
> ```
> 
> Why is `len` the ideal option when you shift the index by -1 position?

If my understanding is right, The perm parameter must be a valid permutation of the indices of the subsystems. This means it should be a rearrangement of the numbers from 0 to 𝑛 -1, where n is the number of subsystems.

If you have a system with 4 subsystems, valid permutations of the indices [0, 1, 2, 3] could be:

```
[0, 1, 2, 3] (identity permutation, no change)
[3, 2, 1, 0] (reverse order)
[1, 0, 3, 2] (swap first two and last two)
[2, 3, 0, 1] (rotate left)

```
so [7,8,9,10] will not be a valid perm list.
Next, the dim parameter should be based on the number of subsystems, doing `max(perm)` would get us 3 while this is a system with 4 subsystems, len() will give us the correct count of the system, earlier when we start perm list with a start index of 1 doing max() got us the right count of subsystems `max([1,2,3,4] ) = 4` but now that the same starts with 0 instead of 1 ` max([0,1,2,3])=3 ` which would be incorrect, So instead of adding +1 after max(),  I used len ()

Reply:
Let's just leave it as `max+1` then. I have a use-case scenario where I think the len option will give the incorrect result. However, I don't think I will be able to test it out over the next few days.

Reply:
Sounds good will update the code to max +1